### DISCUSSION TITLE
S3 Select Performance for CSV

### DISCUSSION THREAD

**Discussion by rajpayagond** [minio/minio#13264] [Q&A] [answered]
Hi,

I tried the S3 Select test as per the demo done in https://www.youtube.com/watch?v=eWbO3D7JM-4&t=955s on minio master.
I see results that are not inline with the demo results.

MinIO Server machine configuration used for test:
Dell R740xd	376GB	2 x Intel(R) Xeon(R) Gold 6252 CPU @ 2.10GHz	2  CPUs x 22 Cores
Skylake architecture with AVX512 SIMD instructions capability
4 NVMe devices 3.84 TB each

Dataset used:
parking-citations.csv with 26,155,449 rows

I tried S3 Select with both mc client and boto3 client, and observe similar results

$ time ./mc sql --query "select count(*) from S3Object where Make='FERR'" myminio/testbuc-1/parking-citations.csv
3998

real    0m32.179s
user    0m0.058s
sys     0m0.058s

I ran the same test run with boto3 client
$ time python3 minio-perf-select-csv.py
3998

real    0m29.312s
user    0m0.174s
sys     0m0.032s

I see the select/simd changes haven't been merged for CSV yet. 
Is there a way to use the select/simd optimizations for CSV?

Thanks,
Raj

**Comment by klauspost** [comment, marked as answer]
@rajpayagond I have experimentally added the simdcsv support to MinIO. It did not provide any speed-up and added a lot of complexity so it was abandoned.

**Reply by somnathr** [reply]
@klauspost thanks for the response, so, the numbers shown by rajpayagond is kind of inline with expected latency of Minio on S3-select then ?

**Reply by harshavardhana** [reply]
> @klauspost thanks for the response, so, the numbers shown by rajpayagond is kind of inline with expected latency of Minio on S3-select then ?

It is as good as the CSV parser.

**Reply by harshavardhana** [reply]
You are better off downloading the entire object locally and run the queries locally it will be way faster with spark.

**Reply by somnathr** [reply]
Thanks Harsha, but, that will kill network BW. Anyways, we were surprised by the difference we are getting and the video was claiming and that's why wanted to check with you guys..

**Reply by harshavardhana** [reply]
> Thanks Harsha, but, that will kill network BW. Anyways, we were surprised by the difference we are getting and the video was claiming and that's why wanted to check with you guys..

CSV datasets can have varied results based on the number of columns keywords etc, especially your query is doing `count()` which is bounded by the number of lines of CSV entries you have. There is nothing we can do magic to make it faster. @somnathr

**Reply by klauspost** [reply]
In this particular case it is evaluating the WHERE clause that is the bottleneck. Either way, SIMD does not change what

FWIW execution time is ~13s on AMD 3950x.
